master/internal/api_trials.go

```diff
@@ -368,6 +370,292 @@ func (a *apiServer) GetTrialCheckpoints(
 	return resp, a.paginate(&resp.Pagination, &resp.Checkpoints, req.Offset, req.Limit)
 }
 
+func checkTrialFiltersEmpty(f *apiv1.TrialFilters) error {
+	emptyFilters := status.Errorf(
+		codes.InvalidArgument,
+		"at least one filter required",
+	)
+
+	if f == nil {
+		return emptyFilters
+	}
+
+	filtersLength := len(f.ExperimentIds) +
+		len(f.ProjectIds) +
+		len(f.WorkspaceIds) +
+		len(f.ValidationMetrics) +
+		len(f.TrainingMetrics) +
+		len(f.Hparams) +
+		len(f.Searcher) +
+		len(f.UserIds) +
+		len(f.Tags) +
+		len(f.States) +
+		len(f.SearcherMetric)
+
+	if filtersLength == 0 && 
+		f.RankWithinExp == nil &&
+	 	f.StartTime == nil && f.EndTime == nil && 
+	 	f.SearcherMetricValue == nil {
+			return emptyFilters
+	}
+	return nil
+}
+
+func (a *apiServer) QueryTrials(ctx context.Context, 
+	req *apiv1.QueryTrialsRequest) (*apiv1.QueryTrialsResponse, error) {
+	err := checkTrialFiltersEmpty(req.Filters)
+	if err != nil {
+		return nil, fmt.Errorf("error querying tags for trials %w", err)
+	}
+
+	trials := []db.TrialsAugmented{}
+
+	q := db.Bun().NewSelect().Model(&trials)
+	q, err = a.m.db.FilterTrials(q, req.Filters, true)
+
+	if err != nil {
+		return nil, fmt.Errorf("error querying for trials %w", err)
+	}
+
+	orderColumn := "trial_id"
+	orderDirection := db.SortDirectionAsc
+	if req.Sorter != nil {
+		orderColumn, err = a.m.db.TrialsColumnForNamespace(req.Sorter.Namespace, req.Sorter.Field)
+		if err != nil {
+			return nil, fmt.Errorf("error querying for trials, bad order by column %w", err)
+		}
+	}
+
+	if req.Limit == 0 {
+		req.Limit = 10
+	}
+
+	q = db.PaginateBun(
+		q,
+		orderColumn,
+		orderDirection,
+		int(req.Offset),
+		int(req.Limit),
+	)
+
+	err = q.Scan(context.TODO())
+
+	if err != nil {
+		return nil, fmt.Errorf("error querying for trials %w", err)
+	}
+
+	resp := apiv1.QueryTrialsResponse{Trials: []*apiv1.AugmentedTrial{}}
+
+	for _, trial := range trials {
+		resp.Trials = append(resp.Trials, trial.Proto())
+	}
+
+	return &resp, nil
+}
+
+func (a *apiServer) PatchTrials(ctx context.Context, 
+	req *apiv1.PatchTrialsRequest) (*apiv1.PatchTrialsResponse, error) {
+	_, _, err := grpcutil.GetUser(ctx, a.m.db, &a.m.config.InternalConfig.ExternalSessions)
+	if err != nil {
+		return nil, fmt.Errorf("couldnt patch trials %w", err)
+	}
+
+	// check user is authorized for modifying project? after RBAC?
+	// in that case we will want to make sure len(req.Filters.ProjectID) == 1
+	// right now only option is adding/removing tags, pretty low stakes
+
+	q := db.Bun().NewUpdate().Table("trials")
+	q, err = a.m.db.ApplyTrialPatch(q, req.Patch)
```

Comment:
can we move this line inside `ApplyTrialPatch` similar to the previous comment? maybe the function is not "applying" anymore, and could be `BuildTrialPatchQuery`

Reply:
ok, by this line you mean `q := db.Bun().NewUpdate().Table("trials")`, right?

Reply:
yes